data-prepper-plugins/aggregate-processor/src/main/java/com/amazon/dataprepper/plugins/processor/aggregate/GroupStateManager.java

```diff
@@ -0,0 +1,41 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package com.amazon.dataprepper.plugins.processor.aggregate;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+
+import com.google.common.collect.Maps;
+
+class GroupStateManager {
+
+    private Map<Object, Map<Object, Object>> allGroupStates = Maps.newConcurrentMap();
+
+    GroupStateManager() {
+
+    }
+
+    Map<Object, Object> getGroupState(final Map<Object, Object> identificationKeysHash) {
```

Comment:
I'm of the opinion that the caller doesn't really care that this is a `Map<Object, Object>`. I can accept the PR without this change, but I'd probably create a new class to abstract this.


```
public static class IdentificationHash {
  private final Map<Object, Object> hash;

  @Override
  public equals(Object other) {
    return this.hash.equals(other);
  }

  @Override
  public int hashCode() {
    return hash.hashCode();
  }
}
```

Then:

```
IdentificationHash getGroupState(...) {
  ...
  return new IdentificationHash(identificationKeysHash);
}
```


You could change how you hash without affecting clients.

Reply:
Why is the class `static`? Is the idea that it is nested in `AggregateIdentificationKeysHasher`?

Reply:
I was thinking it would be nested under `AggregateIdentificationKeysHasher`. It could also be a normal class and un-nested. Either way is fine.

Reply:
I like the idea of abstracting out `IdentificationHash`, but I don't believe `getGroupState` should return an `IdentificationKeyHash`.

Reply:
Soon I plan on creating a `GroupState` class that will be returned in `getGroupState`

Reply:
I'm sorry, that was a copy-and-paste mistake on my part because my comment is here on `getGroupState()`. The `createIdentificationKeyHashFromEvent` method would return an `IdentificationHash`. And then you pass it into `getGroupState`.

I'll update my original comment.

Reply:
I am strongly for a group state class. It's very difficult to read the code as is with all the `Map<Object, Object>` types. I like the suggestion.

Reply:
The thing to consider here is whether a `Map<Object, Object>`, or a `GroupState` is passed to AggregateAction `handleEvent` and `concludeGroup`. Right now, `GroupState` will just contain this Map, but in the future it will contain other things like the windowDuration, readWriteLock, etc. that we definitely don't want `AggregateAction` to have access to, but will probably want `AggregateProcessor` to have access to.

Reply:
I am leaning towards just giving the direct `Map<Object, Object`

Reply:
> I am strongly for a group state class.

@cmanning09 , The discussion here is about is an "Identification Hash" class, not the group state.

Reply:
@graytaylor0 ,

> Soon I plan on creating a GroupState class that will be returned in getGroupState


I very much like the idea here. But, I think the name should be something else. I'd like to keep "group state" as the name of the map which `AggregateProcessor` provides to `AggregateAction`s.

I'd like to use a different term for the object which holds all of this information. Right now, GroupMetadata comes to mind. e.g.

```
public class GroupMetadata {
  private Map<Object, Object> groupState;
  private Instant windowStart;
}
```

Reply:
Yes, the `Map` isn't Metadata. I believe that we should continue to call the `Map` the "group state."

What I'm trying to avoid is calling two different things "group state." Let's call the `Map` "group state" and any class which holds the "group state" some other name.

Reply:
@dlvenable 

> > I am strongly for a group state class.
> 
> @cmanning09 , The discussion here is about is an "Identification Hash" class, not the group state.

I should have quoted the line I was referring to:

> Soon I plan on creating a GroupState class that will be returned in getGroupState